File: 1763-longest-nice-substring/1763-longest-nice-substring.cs
public class Solution {
    public string LongestNiceSubstring(string s) {
        int n = s.Length;
        int maxLength = 0; 
        int maxStart = 0; int maxEnd = 0;

        for (int l = 0; l < n; l++) {
            int r = l + 1;
            while (r <= n) {
                string substr = s.Substring(l, r - l);
                if (IsNice(substr) && substr.Length > maxLength) {
                    maxLength = substr.Length;
                    maxStart = l;
                    maxEnd = r;
                }
                r++;
            }
        }

        return s.Substring(maxStart, maxEnd - maxStart);
    }

    // Helper method
    private bool IsNice(string s) {
        var unique = new HashSet<char>(s);
        foreach (char ch in unique) {
            if (!unique.Contains(char.ToLower(ch)) || !unique.Contains(char.ToUpper(ch))) {
                return false;
            }
        }
        return true;
    }
}

File: 1763-longest-nice-substring/1763-longest-nice-substring.py
class Solution:
    def longestNiceSubstring(self, s: str) -> str:
        def is_nice(sub):
            char_count = {}
            for char in sub:
                char_count[char] = char_count.get(char, 0) + 1

            for char in char_count:
                if char.islower() and char.upper() not in char_count:
                    return False
                if char.isupper() and char.lower() not in char_count:
                    return False

            return True
    
        n = len(s)
        max_length = 0
        max_start = 0
        max_end = 0

        for l in range(n):
            r = l + 1
            while r <= n:
                substring = s[l:r]
                if is_nice(substring) and len(substring) > max_length:
                    max_length = len(substring)
                    max_start = l
                    max_end = r
                r += 1

        return s[max_start:max_end]
    
 








        